common/faults/faults.c

```diff
@@ -0,0 +1,461 @@
+	fault_message_t *currMessage = &messageArray[GET_IDX(recievedMessage.f_ID)];
+	handleCallbacks(recievedMessage, currMessage);
+}
+void fault_sync_l4_testing_CALLBACK(CanParsedData_t *msg_header_a) {
+	fault_message_t recievedMessage = {msg_header_a->fault_sync_l4_testing.latched, msg_header_a->fault_sync_l4_testing.idx};
+	fault_message_t *currMessage = &messageArray[GET_IDX(recievedMessage.f_ID)];
+	handleCallbacks(recievedMessage, currMessage);
+}
+//END AUTO RECIEVE FUNCTIONS
+
+//Force faults from daq
+void set_fault_CALLBACK(CanParsedData_t *msg_header_a) {
+    if (GET_OWNER(msg_header_a->set_fault.id) == currentMCU) {
+        forceFault(msg_header_a->set_fault.id, msg_header_a->set_fault.value);
+    }
+}
+
+//Return control back to this mcu from daq
+void return_fault_control_CALLBACK(CanParsedData_t *msg_header_a) {
+    if (GET_OWNER(msg_header_a->set_fault.id) == currentMCU) {
+        unForce(msg_header_a->set_fault.id);
+    }
+}
+
+
+//Updates faults owned by current mcu
+void updateFaults() {
+    uint16_t idx = ownedidx;
+    fault_attributes_t *fault;
+    do {
+        fault = &faultArray[idx];
+        //Fault is showing up as latched
+        if (((fault->tempLatch) && !(fault->message->latched))) {
+            //Has latching period ended
+            fault->message->latched = (fault->time_since_latch++ >= faultLatchTime[idx]) ? true : false;
+            if (fault->message->latched) {
+                currCount++;
+                switch(fault->priority) {
+                    case INFO:
+                        infoCount++;
+                        break;
+                    case WARNING:
+                        warnCount++;
+                        break;
+                    case CRITICAL:
+                        critCount++;
+                        break;
+                }
+                txFaultSpecific(idx);
+            }
+        }
+        //Fault is showing up as unlatched
+        else if (!(fault->tempLatch) && (fault->message->latched)) {
+            //Make sure unlatch period has elapsed
+            fault->message->latched = (fault->time_since_latch++ >= faultULatchTime[idx]) ? false : true;
+            if (!(fault->message->latched)) {
+                currCount--;
+                switch(fault->priority) {
+                    case INFO:
+                        infoCount--;
+                        break;
+                    case WARNING:
+                        warnCount--;
+                        break;
+                    case CRITICAL:
+                        critCount--;
+                        break;
+                }
+                txFaultSpecific(idx);
+            }
+        }
+        else {
+            //Account for potential noise during the latching process
+            if (fault->time_since_latch > 0 && fault->bounces < 3) {
+                fault->time_since_latch++;
+                fault->bounces++;
+                fault->message->latched = 1;
```

Comment:
Only want fault to latch if it's been >= latch time
The desired functionality is that if the fault is unlatched, but blipping on/off for >= latch time, it should latch

Reply:
Apologies about this, I should have read this over. Here is my idea: while <= 3, the value will be ignored, and bounces will be incremented. I chose 3, but that can be increased. The goal with if and else if is to allow for bounces, but if there are an unreasonable amount of bounces, then something should be done about it. This is why I latched the fault in the else if statement. I should have looked it over and noticed that I copy+pasted and forgot to delete the latching code. I have deleted this.

Reply:
Also, it could be worth making the number of bounces dependent on the latching time. If the latch time is like 2 seconds versus 100 ms, you should accept more bounces. Maybe like if the fault is on > 60% of the time